rts/motoko-rts/src/stream.rs

```diff
@@ -0,0 +1,184 @@
+}
+
+impl Stream {
+    #[inline]
+    pub unsafe fn cache_addr(self: *const Self) -> *const u8 {
+        self.add(1) as *const u8 // skip closure header
+    }
+
+    #[inline]
+    pub unsafe fn as_blob_mut(self: *mut Self) -> *mut Blob {
+        self as *mut Blob
+    }
+
+    /// make sure that the cache is empty
+    fn flush(self: *mut Self) {
+        unsafe {
+            if (*self).filled > INITIAL_STREAM_FILLED {
+                ((*self).outputter)(
+                    self,
+                    self.cache_addr(),
+                    (*self).filled - INITIAL_STREAM_FILLED,
+                );
+                (*self).filled = INITIAL_STREAM_FILLED
+            }
+        }
+    }
+
+    fn no_backing_store(self: *mut Self, _ptr: *const u8, _n: Bytes<u32>) {
+        assert!(false)
+    }
+
+    #[cfg(feature = "ic")]
+    fn send_to_stable(self: *mut Self, ptr: *const u8, n: Bytes<u32>) {
+        unsafe {
+            let next_ptr64 = (*self).ptr64 + n.as_u32() as u64;
+            stable64_write_moc((*self).ptr64, ptr as u64, n.as_u32() as u64);
+            (*self).ptr64 = next_ptr64
+        }
+    }
+
+    #[cfg(feature = "ic")]
+    /// Sets up the bottleneck routine to output towards a range of stable memory
+    /// Note: assumes that the entire byte range is writable
+    #[export_name = "stream_stable_dest"]
+    pub fn setup_stable_dest(self: *mut Self, start: u64, limit: u64) {
+        unsafe {
+            (*self).ptr64 = start;
+            (*self).limit64 = limit;
+            (*self).outputter = Self::send_to_stable;
+        }
+    }
+
+    /// Ingest a number of bytes into the stream.
+    #[export_name = "stream_write"]
+    pub fn cache_bytes(self: *mut Self, ptr: *const u8, n: Bytes<u32>) {
+        unsafe {
+            if (*self).limit64 != 0 && n > STREAM_CHUNK_SIZE
+                || (*self).filled + n > (*self).header.len
+            {
+                self.flush();
+                ((*self).outputter)(self, ptr, n);
+            } else {
+                let dest = self
+                    .as_blob_mut()
+                    .payload_addr()
+                    .add((*self).filled.as_usize());
+                (*self).filled += n;
+                assert!((*self).filled <= (*self).header.len);
+                memcpy_bytes(dest as usize, ptr as usize, n);
+            }
+        }
+    }
+
+    /// Ingest a single byte into the stream.
+    #[export_name = "stream_write_byte"]
+    pub fn cache_byte(self: *mut Self, byte: u8) {
+        unsafe {
+            if (*self).filled >= (*self).header.len {
+                self.flush()
+            }
+            self.as_blob_mut().set((*self).filled.as_u32(), byte);
+            (*self).filled += Bytes(1)
+        }
+    }
+
+    /// Return a pointer to a reserved area of the cache and advance the
+    /// fill indicator beyond it.
+    #[export_name = "stream_reserve"]
+    pub fn reserve(self: *mut Self, bytes: Bytes<u32>) -> *mut u8 {
```

Comment:
What happens if you reserve some space, returning its pointer, but the next write flushes the cache. Wouldn't it be bad to now use that pointer? Is this guaranteed to never happen?